File: java/src/bfs/BreadthFirstSearchTest.java
package bfs;

import org.junit.Assert;
import org.junit.Test;

import java.util.List;

public class BreadthFirstSearchTest {
	@Test
	public void testBfs() {
		var graph = List.of(
			List.of(1, 3),
			List.of(2,4),
			List.of(5),
			List.of(4,6),
			List.of(5,7),
			List.of(8),
			List.of(7),
			List.of(8),
			List.of(2)
		);
		var expected = List.of(0, 1, 3, 2, 4, 6, 5, 7, 8);

		var actual = BreadthFirstSearch.bfs(graph, 0);
		if (!actual.equals(expected)) {
			Assert.fail(String.format("dfs(%s, 0):\nwant %s\n got %s", graph, expected, actual));
		}
	}

	@Test
	public void testRandomCases() {
		for (int i = 0; i < 10_000; i++) {
			var td = new randomTestData();

			var expected = td.order;
			var actual = BreadthFirstSearch.bfs(td.graph, td.start);

			if (!expected.equals(actual)) {
				Assert.fail(String.format(
					"bfs(%s, %s):\nwant %s\n got %s",
					td.graph, td.start, expected, actual
				));
			}
		}
	}
}

File: python/bfs/breadth_first_search_test.py
from breadth_first_search import bfs
from random_test_data import RandomTestData


def test_bfs():
	graph = [
		[1, 3],
		[2, 4],
		[5],
		[4, 6],
		[5, 7],
		[8],
		[7],
		[8],
		[2],
	]

	want = [0, 1, 3, 2, 4, 6, 5, 7, 8]

	got = bfs(graph, 0)

	if got != want:
		assert False, f'bfs({graph}, 0):\nwant {want}\n got {got}'


def test_random_cases():
	for _ in range(10_000):
		td = RandomTestData()

		want = td.order
		got = bfs(td.graph, td.start)

		if got != want:
			assert False, f'bfs({td.graph}, {td.start}):\nwant {want}\n got {got}'
